memory-db/src/lib.rs

```diff
@@ -350,16 +377,25 @@ where
 	/// }
 	/// ```
 	pub fn clear(&mut self) {
+		self.malloc_callback.on_clear();
 		self.data.clear();
 	}
 
 	/// Purge all zero-referenced data from the database.
 	pub fn purge(&mut self) {
-		self.data.retain(|_, &mut (_, rc)| rc != 0);
+		let malloc_callback = &mut self.malloc_callback;
+		self.data.retain(|_, (v, rc)| {
+			let to_retain = *rc != 0;
+			if !to_retain {
+				malloc_callback.on_value_removed(v);
+			}
+			to_retain
+		});
 	}
 
 	/// Return the internal map of hashes to data, clearing the current state.
 	pub fn drain(&mut self) -> HashMap<KF::Key, (T, i32)> {
+		self.malloc_callback.on_clear();
```

Comment:
Maybe the callback should be called `on_drain`?

Reply:
I prefer `on_clear` in this case, since it would apply to a `clear` function the same way.